crates/bevy_ecs/src/system/query.rs

```diff
@@ -565,6 +638,53 @@ impl<'w, 's, Q: WorldQuery, F: WorldQuery> Query<'w, 's, Q, F> {
         };
     }
 
+    /// Calls a closure on each result of [`Query`] where the entities match.
+    /// # Examples
+    ///
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// #[derive(Component)]
+    /// struct Counter {
+    ///     value: i32
+    /// }
+    ///
+    /// #[derive(Component)]
+    /// struct Friends {
+    ///     list: Vec<Entity>,
+    /// }
+    ///
+    /// fn system(
+    ///     friends_query: Query<&Friends>,
+    ///     mut counter_query: Query<&mut Counter>,
+    /// ) {
+    ///     for friends in &friends_query {
+    ///         counter_query.many_for_each_mut(&friends.list, |mut counter| {
+    ///             println!("Friend's counter: {:?}", counter.value);
+    ///             counter.value += 1;
+    ///         });
+    ///     }
+    /// }
+    /// # bevy_ecs::system::assert_is_system(system);
+    /// ```
+    pub fn many_for_each_mut<E: Borrow<Entity>, EntityList: IntoIterator<Item = E>>(
```

Comment:
I just ensured that data isn't allowed to escape the closure (which is definitely what we want). To ensure we never regress here (ex: by changing this to `QueryItem<'w, Q>`), we should add a new "compile fail" test to `bevy_ecs_compile_fail_tests`. Check out tests like `system_query_get_many_lifetime_safety` for an example.

This is the type of thing that should never be allowed to compile, as it essentially turns this into a mutable iterator:

```rust
let mut results = Vec::new();
query.many_for_each_mut(vec![e, e], |a| {
    results.push(a);
});
```